tests/queries/0_stateless/02983_variant_data_types.sh

```diff
@@ -0,0 +1,312 @@
+    run_select "DateTime"
+    drop_table
+
+    create_table "$1" "DateTime64"
+    $CH_CLIENT -q  "insert into test values (1, '2299-01-01 00:00:00');"
+    # run_select "DateTime64" # Cannot select referring to data type DateTime64 or DateTime64(3)
+    drop_table
+}
+
+function test_json()
+{
+    echo "JSON Types"
+
+    #create_table "$1" "JSON" # Variant accepts the JSON type with exception : "Code: 36. DB::Exception: Positional options are not supported. (BAD_ARGUMENTS)"
+    #CH_CLIENT -q  "insert into test values (1, '{"a": 1, "b": { "c": 2, "d": [1, 2, 3] }}')";
+    #run_select "JSON" # select fails with the expected error "unexpectedly has dynamic columns. (LOGICAL_ERROR)"
+    #drop_table
+}
+
+function test_uuid()
+{
+    echo "UUID Types"
+
+    create_table "$1" "UUID"
+    $CH_CLIENT -q  "insert into test select 1, toUUID('61f0c404-5cb3-11e7-907b-a6006ad3dba0');"
+    run_select "UUID"
+    drop_table
+}
+
+function test_low_cardinality_types()
+{
+    echo "Low Cardinality Types"
+
+    create_table "$1" "LowCardinality(String)"
+    $CH_CLIENT -q  "insert into test select number, 'str_' || toString(number)::LowCardinality(String) from numbers(10);"
+    run_select "LowCardinality(String)"
+    drop_table
+
+    #create_table "$1" "Enum"
+    #$CH_CLIENT -q  "insert into test run_select number, number % 2, from numbers(5);"
+    #run_select "Enum"
+    #drop_table
+}
+
+function test_arrays()
+{
+    echo "Array Types"
+
+    create_table "$1" "Array(UInt64)"
+    $CH_CLIENT -q  "insert into test select 1, range(number) from numbers(5);"
+    run_select "Array(UInt64)"
+    drop_table
+
+    create_table "$1" "Array(Variant(UInt64, String))"
+    $CH_CLIENT -q  "insert into test select 1, (number % 2) ? 'str_' || toString(number) : array(number, 'str__' || toString(number)) from numbers(5);"
+    #run_select "Array(Variant(UInt64, String))" # select doesn't work. Probably because of multiple parentheses in type name
+    drop_table
+}
+
+function test_maps()
+{
+    echo "Map Types"
+
+    create_table "$1" "Map(String, String)"
+    $CH_CLIENT -q  "insert into test select 1, (number % 2 ? toString(number) : map('a', toString(number), 'b', 'str_' || toString(number))) from numbers(5);"
+    run_select "Map(String, String)"
+    drop_table
+    
+    create_table "$1" "Map(String, Variant(UInt8, String))"
+    $CH_CLIENT -q  "insert into test select 1, (map('a', toString(number), 'b', number % 2))::Map(String, Variant(UInt8, String)) from numbers(5);"
+    #run_select "Map(String, Variant(UInt8, String))"  # select doesn't work. Probably because of multiple parentheses in type name
+    drop_table
+}
+
+function test_nested_types()
+{
+    echo "Nested Types"
+
+    create_table "$1" "Nested(i UInt64, s String)" # Create table allowed with Nested type as valid type
+    #$CH_CLIENT -q  "insert into test run_select 1, (number % 2 ? toString(number) : map('a', toString(number), 'b', 'str_' || toString(number) ) from numbers(5);"
```

Comment:
You can insert into table with nested with smth like this:
```sql
insert into test select 1, number % 2 ? toString('str_' || toString(number)) : [(number, 'str_' || toString(number))]::Nested(i UInt64, s String) from numbers(5)
```

Reply:
Thank you. Updated.